Remove overly strict assert, add test

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #94159
* #94152
* #94150
* __->__ #94151
* #94285



cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

voznesenskym committed:
Remove overly strict assert, add test

[ghstack-poisoned]

ezyang reviewed (changes requested):
This looks wrong. You need to check the types are compatible.

```
>>> [1] == (1,)
False
```

voznesenskym commented:
> This looks wrong. You need to check the types are compatible.
> 
> 
> 
> ```
> 
> >>> [1] == (1,)
> 
> False
> 
> ```

Good catch. I'll add a carve out for size.

voznesenskym committed:
Update on "Remove overly strict assert, add test"




cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Remove overly strict assert, add test"




cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on test/dynamo/test_misc.py:
```diff
@@ -245,6 +245,29 @@ def compare_shapes(a, b, to_list):
+        )(compare_shapes)
+        opt_fn(torch.randn([3, 4]))
+        opt_fn(torch.randn([4, 3]))
+        self.assertEqual(guard_failure.reason, "a.size()[0] == 3")
```
Not sure what this test has to do with the PR

voznesenskym replied:
Nothing, will give its own PR

voznesenskym replied:
eh too much hassle, I figured you dont mind

ezyang commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1009,6 +1012,13 @@ def _unimplemented():
+        # (3, 3) == x.size() # True
+        if isinstance(left, (SizeVariable, TupleVariable)) and isinstance(right, (TupleVariable, SizeVariable)):
+            return BaseListVariable.generic_list_compare(left, tx, op, right)
+
```
This isn't very generic but it's probably fine

ezyang reviewed (approved):
Do you want a test for tuple list comparison?

voznesenskym committed:
Update on "Remove overly strict assert, add test"




cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Remove overly strict assert, add test"




cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Remove overly strict assert, add test"




cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
And this produces another inductor error reliably
```
Traceback (most recent call last):
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_inductor/graph.py", line 321, in call_function
    out = lowerings[target](*args, **kwargs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_inductor/lowering.py", line 224, in wrapped
    return decomp_fn(*args, **kwargs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_inductor/lowering.py", line 777, in split
    for size in sizes:
TypeError: 'Integer' object is not iterable
ERROR:common:inductor raised LoweringException: TypeError: 'Integer' object is not iterable
  target: aten.split.Tensor
  args[0]: TensorBox(StorageBox(
    ConcatKernel(name='buf2', layout=FixedLayout('cuda', torch.float16, size=[s0, s1, 512], stride=[512*s1, 512, 1]), inputs=[ComputedBuffer(name='buf0', layout=AliasedLayout('cuda', torch.float16, size=[s0, s1, 256], stride=[512*s1, 512, 1]), data=Pointwise(
      'cuda',
      torch.float16,
      tmp0 = load(arg77_1, i2 + 256 * i1 + 256 * i0 * s1)
      return tmp0
      ,
      ranges=[s0, s1, 256],
      origins={cat}
    )), ComputedBuffer(name='buf1', layout=AliasedLayout('cuda', torch.float16, size=[s0, s1, 256], stride=[512*s1, 512, 1]), data=Pointwise(
      'cuda',
      torch.float16,
      tmp0 = load(arg77_1, i2 + 256 * i1 + 256 * i0 * s1)
      return tmp0
      ,
      ranges=[s0, s1, 256],
      origins={cat}
    ))])
  ))
  args[1]: 256
  args[2]: -1

While executing %split : [#users=2] = call_function[target=torch.ops.aten.split.Tensor](args = (%cat, %floordiv, -1), kwargs = {})
Original traceback:
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/transformers/models/reformer/modeling_reformer.py", line 1615, in forward
    hidden_states, attn_output = torch.chunk(hidden_states, 2, dim=-1)
 |   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/transformers/models/reformer/modeling_reformer.py", line 1733, in forward
    hidden_states = _ReversibleFunction.apply(
 |   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/transformers/models/reformer/modeling_reformer.py", line 2107, in <graph break in forward>
    encoder_outputs = self.encoder(


Set torch._dynamo.config.verbose=True for more information


You can suppress this exception and fall back to eager by setting:
    torch._dynamo.config.suppress_errors = True
Traceback (most recent call last):
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_inductor/graph.py", line 321, in call_function
    out = lowerings[target](*args, **kwargs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_inductor/lowering.py", line 224, in wrapped
    return decomp_fn(*args, **kwargs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_inductor/lowering.py", line 777, in split
    for size in sizes:
TypeError: 'Integer' object is not iterable

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/output_graph.py", line 670, in call_user_compiler
    compiled_fn = compiler_fn(gm, self.fake_example_inputs())
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/debug_utils.py", line 1055, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/backends/inductor.py", line 9, in inductor
    return compile_fx(*args, **kwargs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_inductor/compile_fx.py", line 426, in compile_fx
    return aot_autograd(
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/backends/common.py", line 48, in compiler_fn
    cg = aot_module_simplified(gm, example_inputs, **kwargs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 2487, in aot_module_simplified
    compiled_fn = create_aot_dispatcher_function(
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/utils.py", line 163, in time_wrapper
    r = func(*args, **kwargs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 2184, in create_aot_dispatcher_function
    compiled_fn = compiler_fn(flat_fn, fake_flat_args, aot_config)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 1415, in aot_wrapper_dedupe
    return compiler_fn(flat_fn, leaf_flat_args, aot_config)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 1065, in aot_dispatch_base
    compiled_fw = aot_config.fw_compiler(fw_module, flat_args)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/utils.py", line 163, in time_wrapper
    r = func(*args, **kwargs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_inductor/compile_fx.py", line 401, in fw_compiler
    return inner_compile(
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/debug_utils.py", line 595, in debug_wrapper
    compiled_fn = compiler_fn(gm, example_inputs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_inductor/debug.py", line 239, in inner
    return fn(*args, **kwargs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_inductor/compile_fx.py", line 154, in compile_fx_inner
    graph.run(*example_inputs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/utils.py", line 163, in time_wrapper
    r = func(*args, **kwargs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_inductor/graph.py", line 182, in run
    return super().run(*args)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/fx/interpreter.py", line 136, in run
    self.env[node] = self.run_node(node)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_inductor/graph.py", line 397, in run_node
    result = super().run_node(n)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/fx/interpreter.py", line 177, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_inductor/graph.py", line 325, in call_function
    raise LoweringException(e, target, args, kwargs) from e
torch._inductor.exc.LoweringException: TypeError: 'Integer' object is not iterable
  target: aten.split.Tensor
  args[0]: TensorBox(StorageBox(
    ConcatKernel(name='buf2', layout=FixedLayout('cuda', torch.float16, size=[s0, s1, 512], stride=[512*s1, 512, 1]), inputs=[ComputedBuffer(name='buf0', layout=AliasedLayout('cuda', torch.float16, size=[s0, s1, 256], stride=[512*s1, 512, 1]), data=Pointwise(
      'cuda',
      torch.float16,
      tmp0 = load(arg77_1, i2 + 256 * i1 + 256 * i0 * s1)
      return tmp0
      ,
      ranges=[s0, s1, 256],
      origins={cat}
    )), ComputedBuffer(name='buf1', layout=AliasedLayout('cuda', torch.float16, size=[s0, s1, 256], stride=[512*s1, 512, 1]), data=Pointwise(
      'cuda',
      torch.float16,
      tmp0 = load(arg77_1, i2 + 256 * i1 + 256 * i0 * s1)
      return tmp0
      ,
      ranges=[s0, s1, 256],
      origins={cat}
    ))])
  ))
  args[1]: 256
  args[2]: -1

While executing %split : [#users=2] = call_function[target=torch.ops.aten.split.Tensor](args = (%cat, %floordiv, -1), kwargs = {})
Original traceback:
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/transformers/models/reformer/modeling_reformer.py", line 1615, in forward
    hidden_states, attn_output = torch.chunk(hidden_states, 2, dim=-1)
 |   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/transformers/models/reformer/modeling_reformer.py", line 1733, in forward
    hidden_states = _ReversibleFunction.apply(
 |   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/transformers/models/reformer/modeling_reformer.py", line 2107, in <graph break in forward>
    encoder_outputs = self.encoder(


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/benchmarks/dynamo/common.py", line 1244, in check_accuracy
    new_result = optimized_model_iter_fn(model_copy, example_inputs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 209, in _fn
    return fn(*args, **kwargs)
  File "/var/lib/jenkins/workspace/benchmarks/dynamo/common.py", line 1121, in run_n_iterations
    self.model_iter_fn(mod, inputs, collect_outputs=False)
  File "/var/lib/jenkins/workspace/benchmarks/dynamo/torchbench.py", line 362, in forward_pass
    return mod(*inputs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/transformers/models/reformer/modeling_reformer.py", line 2397, in forward
    reformer_outputs = self.reformer(
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/transformers/models/reformer/modeling_reformer.py", line [2063](https://github.com/pytorch/pytorch/actions/runs/4107469412/jobs/7087445619#step:11:2064), in forward
    least_common_mult_chunk_length = _get_least_common_mult_chunk_len(self.config)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/transformers/models/reformer/modeling_reformer.py", line 2100, in <graph break in forward>
    embedding_output = self.embeddings(
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 330, in catch_errors
    return callback(frame, cache_size, hooks)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 404, in _convert_frame
    result = inner_convert(frame, cache_size, hooks)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 104, in _fn
    return fn(*args, **kwargs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 262, in _convert_frame_assert
    return _compile(
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/utils.py", line 163, in time_wrapper
    r = func(*args, **kwargs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 324, in _compile
    out_code = transform_code_object(code, transform)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/bytecode_transformation.py", line 361, in transform_code_object
    transformations(instructions, code_options)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 311, in transform
    tracer.run()
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 1683, in run
    super().run()
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 569, in run
    and self.step()
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 532, in step
    getattr(self, inst.opname)(inst)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 1749, in RETURN_VALUE
    self.output.compile_subgraph(self)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/output_graph.py", line 541, in compile_subgraph
    self.compile_and_call_fx_graph(tx, pass2.graph_output_vars(), root)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/output_graph.py", line 588, in compile_and_call_fx_graph
    compiled_fn = self.call_user_compiler(gm)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/utils.py", line 163, in time_wrapper
    r = func(*args, **kwargs)
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/output_graph.py", line 675, in call_user_compiler
    raise BackendCompilerFailed(self.compiler_fn, e) from e
torch._dynamo.exc.BackendCompilerFailed: inductor raised LoweringException: TypeError: 'Integer' object is not iterable
  target: aten.split.Tensor
  args[0]: TensorBox(StorageBox(
    ConcatKernel(name='buf2', layout=FixedLayout('cuda', torch.float16, size=[s0, s1, 512], stride=[512*s1, 512, 1]), inputs=[ComputedBuffer(name='buf0', layout=AliasedLayout('cuda', torch.float16, size=[s0, s1, 256], stride=[512*s1, 512, 1]), data=Pointwise(
      'cuda',
      torch.float16,
      tmp0 = load(arg77_1, i2 + 256 * i1 + 256 * i0 * s1)
      return tmp0
      ,
      ranges=[s0, s1, 256],
      origins={cat}
    )), ComputedBuffer(name='buf1', layout=AliasedLayout('cuda', torch.float16, size=[s0, s1, 256], stride=[512*s1, 512, 1]), data=Pointwise(
      'cuda',
      torch.float16,
      tmp0 = load(arg77_1, i2 + 256 * i1 + 256 * i0 * s1)
      return tmp0
      ,
      ranges=[s0, s1, 256],
      origins={cat}
    ))])
  ))
  args[1]: 256
  args[2]: -1

While executing %split : [#users=2] = call_function[target=torch.ops.aten.split.Tensor](args = (%cat, %floordiv, -1), kwargs = {})
Original traceback:
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/transformers/models/reformer/modeling_reformer.py", line 1615, in forward
    hidden_states, attn_output = torch.chunk(hidden_states, 2, dim=-1)
 |   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/transformers/models/reformer/modeling_reformer.py", line 1733, in forward
    hidden_states = _ReversibleFunction.apply(
 |   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/transformers/models/reformer/modeling_reformer.py", line 2107, in <graph break in forward>
    encoder_outputs = self.encoder(


Set torch._dynamo.config.verbose=True for more information


You can suppress this exception and fall back to eager by setting:
    torch._dynamo.config.suppress_errors = True

TorchDynamo optimized model failed to run because of following error
FAIL
```

voznesenskym commented:
So why does it think sizes is an int... hmmm

voznesenskym committed:
Update on "Remove overly strict assert, add test"




cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Remove overly strict assert, add test"




cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Remove overly strict assert, add test"




cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Remove overly strict assert, add test"




cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Remove overly strict assert, add test"




cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
@pytorchbot merge
